### ISSUE TITLE
Since Update on 4th Match, 1.22 Unable to Add Consumer Key

### ISSUE THREAD

**Issue by PHOUSTON90** [tprouvot/Salesforce-Inspector-reloaded#328]
After recent update, I see a generate new token banner on the inspector.

I had previously saved my consumer key in the "Data export" option 
However It looks to have been lost on Update and I can't see the Data Export option now to Re-enter it.

It's now un-usable.

![image](https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/162155779/41321f95-e61f-4cd9-aa6b-9a0faf723812)

**Comment by LCh07** [comment]
Hi @tprouvot, 
Started testing these steps in sandbox and early results are positive.
Noticed your comments about default connected app. I'm trying to weigh up whether we want to go down this route since I suspect it would mean uninstall of current one, generate a new Oauth entry and install app from that entry. 
Can you please advise whether  users will only need to click 'Generate new token' to gain access or will they still need to know/ enter the consumer key. I'm thinking of what happens when we refresh a sandbox or give access to new users.

Thanks

@PHOUSTON90

**Comment by LCh07** [comment]
@tprouvot our testing has shown that users who had not previously used the extension in an org get the error message below:
OAUTH_APP_BLOCKED This app is blocked by Admin

We can only overcome this by granting 'Use Any API Client' to user, getting user to connect and then add Consumer key. After that 'Use Any API Client' can be removed and on next login they come in via our Connected App

**Comment by tprouvot** [comment]
Hi @LCh07,

The users that did not saved a consumer key don't need to do it, they can use the default one.
Based on the error "OAUTH_APP_BLOCKED This app is blocked by Admin" I think you missed a part which is the (default) connected app installation and configuration.

https://tprouvot.github.io/Salesforce-Inspector-reloaded/troubleshooting/#unauthorized-or-network-error



To do once by the admin:

1. Generate new token

<img width="509" alt="image" src="https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/35368290/eb20def4-2ef0-42ff-a11c-5b282031cd0c">

2. Go to "Connected Apps OAuth Usage" to install the connected app

<img width="1357" alt="image" src="https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/35368290/e3e37936-da52-468a-b763-365bf48f4125">

3. Configure the app (add profile or pset)

<img width="579" alt="image" src="https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/35368290/c8062996-2403-4f2b-8b4a-32adc303970c">

4. Check on  "Connected Apps OAuth Usage" that the app is not blocked, if it is then unblock it

5. Generate new token

Then the other users should have no error when clicking on "Generate new token" banner

Let me know if it solved your issue

**Comment by scox-workato** [comment]
I am seeing this as well for users that did not have the app install prior to the update. The app is not listed as blocked in "Connected Apps OAuth Usage"

**Comment by scox-workato** [comment]
I've done that. I've followed all instructions. The app works for all users who have not recently installed it. All users are on v1.22.

**Comment by pyao-bwc** [comment]
Hi @scox-workato , I'm seeing the same as you. I reinstalled the app and it doesn't work, but others in my org who haven't reinstalled can use it fine (after we updated the callback url).

**Comment by tprouvot** [comment]
@scox-workato if my understanding is correct you had previously (before 1.22) created a custom connected app and saved its consumer key in data export.

If you did not uninstall the extension, this consumer key should still be saved in your local storage. Have you updated the callback url of this custom connected app? 

If you want to use the default connected app introduced in 1.22, you can delete property ending with "_clientId" in your local storage (see https://tprouvot.github.io/Salesforce-Inspector-reloaded/how-to/#migrate-saved-queries-from-legacy-extension-to-salesforce-inspector-reloaded for accessing local storage) and then generate new token.

**Comment by scox-workato** [comment]
I have reinstalled the extension. There is no property ending in _clientId in my local storage. The concern is that the connected app is working for users that had connected via earlier versions of the extension, but in 1.22 the extension is only usable by users that have not reinstalled or newly installed the update.

**Comment by LCh07** [comment]
@tprouvot 
Is installing the default connected app the only way to get new users to use the extension? What are the other benefits of the default connected app in relation to future updates you may make (want to understand whether default will avoid this type of problems in the future or do we have other ways to do this)?

In a sandbox I've tried deleting our custom app and then attempting to create a new Oauth Entry from which to create the default connected app but this gets stuck (see attached), and can't find property ending in _clientId.

Thanks
[SF Inspector reloaded - Agdev30 - Install default.docx](https://github.com/tprouvot/Salesforce-Inspector-reloaded/files/14507145/SF.Inspector.reloaded.-.Agdev30.-.Install.default.docx)

**Comment by tprouvot** [comment]
@LCh07 No it is not the only way, but it is the easiest since the users won't have to save a consumer key in the extension.
We should not make any updates on the connected app so the only benefit is to make the configuration easier (no connected app creation and configuration and no need to save the consumer key on each user computer).
Deleting the custom connected app won't have any effect on your configuration.

Make sure to inspect the popup page (right click in the popup), then you should see all the localStorage used for the extension and search for _clientId

<img width="434" alt="image" src="https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/35368290/32c6fd96-6463-41a2-9455-1e47366cd6dd">

![image](https://github.com/tprouvot/Salesforce-Inspector-reloaded/assets/35368290/f61457dc-4b3c-4aa5-88a4-c9d94ca35ccf)

**Comment by mehdisfdc** [comment]
Hi all!
Let me add a few additional explanations on top of what Thomas already said:

Before the latest release of the extension, the only way to use it with API Access Control was to create a connected app, add its Consumer Key to the extension settings, and then not forget about refreshing the access token any time it expired.
That was really cumbersome, which is why we introduced a default (and optional) connected app. 
With the new version, once the app is approved (which you do just once _per org_), users don't have anything to do other than clicking the link that shows in the extension.

For this default connected app to work, we had to make a change to the redirect URL used in the OAuth flow, which is why old users of the extensions see a breaking change in orgs with API Access Control enabled. And I'm sincerely sorry for that impact! (But it was the only way to significantly improve the user experience, so overall I think it's worth it).

Now, with the new version, you have 2 ways to use the extensions in orgs with API Access Control:

1) Legacy way, with your own Connected App.
For this to work, you need to:
1.1) Make sure the callback URL allowed in the connected app is 
`chrome-extension://hpijlohoihegkfehhibggnkbjhoemldh/data-export.html` (which is a change compared to the previous release of the extension)
1.2) And, make sure each user of the extension have registered the Consumer Key of your own Connected App in their Inspector Reloaded install (under Options | API Consumer Key). This step is no different that in the previous version of the extension, and is still required if you want to use your own custom app.


2) New way, much simpler for new users: install and use the new Connected App.
Pre-requisite: if you used an older version of Inspector reloaded + your connected app, make sure to clean the entry in local storage. Alternatively, you can register manually the key of the new app under Options | API Consumer Key: `3MVG9HB6vm3GZZR9qrol39RJW_sZZjYV5CZXSWbkdi6dd74gTIUaEcanh7arx9BHhl35WhHW4AlNUY8HtG2hs `
You also need to install and authorize for all relevant users the new "Salesforce Inspector reloaded" Connected App (as with any app you want to use in an org with API Access Control enabled).

**Comment by mehdisfdc** [comment]
@scox-workato If your new users got the "OAUTH_APP_BLOCKED This app is blocked by Admin" error, it most likely means that then haven't entered the Consumer Key of your own Connected App in their Inspector Reloaded install (under Options | API Consumer Key). In that case, the extension defaults to the new default connected app (that you may not have pre-authorized at that point for the impacted users).

**Comment by pyao-bwc** [comment]
@mehdisfdc , @tprouvot , thanks for all of your work and documentation on this app! I truly appreciate it. And it'll be nice to not have to refresh the access token with the latest version.

@scox-workato , I was seeing the same issue as you. I followed tprouvot's [instructions](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/328#issuecomment-1978830135) from above more carefully and it worked. I missed adding the default Salesforce Inspector reloaded connected app to the permission set, and also didn't unblock the connected app when it was blocked.

**Comment by mehdisfdc** [comment]
Thank @pyao-bwc !
And just to make sure I've set the right expectations: the new version does not eliminate the need to refresh the token after it expired. But it makes the process faster and more user friendly.

**Comment by scox-workato** [comment]
@mehdisfdc @tprouvot I do not have any more time to talk through this issue so I'm going to just move on. I too was able to get it working with the default app. However, with a custom app, the clientId was not being set under the correct name no matter how many different things we tried. We set the consumer key, changed the callback url, made sure the app was unblocked, and made sure the users were on v1.22. I read through the docs all FOUR times I was referred to them on this thread. The consumer key for new users was being saved in local storage as null_clientid because with the connected app, the sfdc url was set to NULL. I have to manually update the local storage to have the correct _clientId prefix. 

To reproduce, you could setup a custom connected app with a newly installed salesforce connector reloaded extension and attempt to set the consumer key. @pyao-bwc the default app works for us, but this still is not a resolution to the custom connected app issue. If any of the developers would test a custom app with a newly installed extension (or cleared cache, empty extension local storage) instead of repeatedly referring to the docs, they would see the same issue several of us our seeing, that storing the consumer key and trying to generate a new token does not currently work.

**Comment by mehdisfdc** [comment]
@scox-workato Thank you for insisting!
So, taking a closer look, there is indeed a bug in this version the way the clientId is stored, when entered from the "Data Export" tab. The new "API Consumer Key" field (in the new section tab) works as expected, but this one is not accessible before successful authentication.

@tprouvot  I will log a separate issue (with more details), as it's different than than the issue reported by @PHOUSTON90

**Comment by scox-workato** [comment]
@mehdisfdc Thank you! I appreciate your help. This extension is incredible, we are glad we can continue using it.
